Add TSQL to linguist

beau-witter opened the pull request:
<!--- Briefly describe what you're changing. -->

## Description
<!--- If necessary, go into depth of what this pull request is doing. -->
Adding the TSQL language as identifiable through linguist.

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] **I am associating a language with a new file extension.**
  - [ ] The new extension is used in hundreds of repositories on GitHub.com
    - Search results for each extension:
      <!-- Replace FOOBAR with the new extension, and KEYWORDS with keywords unique to the language. Repeat for each extension added. -->
      - https://github.com/search?utf8=%E2%9C%93&type=Code&ref=searchresults&q=extension%3AFOOBAR+KEYWORDS+NOT+nothack
  - [ ] I have included a real-world usage sample for all extensions added in this PR:
    - Sample source(s):
      - [URL to each sample source, if applicable]
    - Sample license(s):
  - [ ] I have included a change to the heuristics to distinguish my language from others using the same extension.

- [x] **I am adding a new language.**
  - [x] The extension of the new language is used in hundreds of repositories on GitHub.com.
    - Search results for each extension:
      <!-- Replace FOOBAR with the new extension, and KEYWORDS with keywords unique to the language. Repeat for each extension added. -->
      -  https://github.com/search?utf8=%E2%9C%93&type=Code&ref=searchresults&q=extension%3Asql+GO
  - [x] I have included a real-world usage sample for all extensions added in this PR:
    - Sample source(s):
      - [URL to each sample source, if applicable]
    - Sample license(s):
  <!-- Update the Lightshow URLs below to show the grammar in action if you included one. -->
  - [x] I have included a syntax highlighting grammar: https://github-lightshow.herokuapp.com/
  - [x] I have updated the heuristics to distinguish my language from others using the same extension.

- [ ] **I am fixing a misclassified language**
  - [ ] I have included a new sample for the misclassified language:
    - Sample source(s):
      - [URL to each sample source, if applicable]
    - Sample license(s):
  - [ ] I have included a change to the heuristics to distinguish my language from others using the same extension.

- [ ] **I am changing the source of a syntax highlighting grammar**
  <!-- Update the Lightshow URLs below to show the new and old grammars in action. -->
  - Old: https://github-lightshow.herokuapp.com/
  - New: https://github-lightshow.herokuapp.com/
  
- [ ] **I am updating a grammar submodule**
  <!-- That's not necessary, grammar submodules are updated automatically with each new release. -->

- [ ] **I am adding new or changing current functionality**
  <!-- This includes modifying the vendor, documentation, and generated lists. -->
  - [ ] I have added or updated the tests for the new or changed functionality.

beau-witter commented on lib/linguist/heuristics.yml:
```diff
@@ -383,13 +383,17 @@ disambiguations:
   rules:
    # Postgres
   - language: PLpgSQL
-    pattern: '(?i:^\\i\b|AS \$\$|LANGUAGE ''?plpgsql''?|SECURITY (DEFINER|INVOKER)|BEGIN( WORK| TRANSACTION)?;)'
```
This change is partially due to Issue 4386.

beau-witter commented on lib/linguist/heuristics.yml:
```diff
@@ -383,13 +383,17 @@ disambiguations:
     pattern: '(?i:\$\$PLSQL_|XMLTYPE|sysdate|systimestamp|\.nextval|connect by|AUTHID (DEFINER|CURRENT_USER)|constructor\W+function)'
+  # T-SQL
+  - language: TSQL
+    pattern: '(?i:(GO)|(CREATE PROCEDURE)|OUTPUT( INSERTED)|IF|ELSE|CURSOR|FETCH|DEALLOCATE|DECLARE)'
```
Also as part of Issue 4386, I should add `BEGIN TRY` and `BEGIN CATCH` in the pattern.

lildude replied:
Go for it, if you're sure they're specific to TSQL.

beau-witter commented on samples/TSQL/cursor.sql:
```diff
@@ -0,0 +1,12 @@
+END;
+CLOSE Employee_Cursor;
+DEALLOCATE Employee_Cursor;  
+GO
```
All 3 of these were taken from various subsections of this section of Microsoft documentation https://docs.microsoft.com/en-us/sql/t-sql/queries/queries?view=sql-server-2017

lildude replied:
We can't use these samples, even if GitHub is part of Microsoft now, as it's not clear what license these files are covered by.

beau-witter replied:
It looks like this is the license covered by those files. https://github.com/MicrosoftDocs/sql-docs/blob/live/LICENSE

lildude replied:
Good find. I see the code itself is covered by MIT - https://github.com/MicrosoftDocs/sql-docs/blob/live/LICENSE-CODE - so we're good. 🙇

beau-witter commented on vendor/README.md:
```diff
@@ -378,6 +378,7 @@ This is a list of grammars that Linguist selects to provide syntax highlighting
 - **SystemVerilog:** [bitbucket:Clams/sublimesystemverilog](https://bitbucket.org/Clams/sublimesystemverilog)
 - **TLA:** [agentultra/TLAGrammar](https://github.com/agentultra/TLAGrammar)
 - **TOML:** [textmate/toml.tmbundle](https://github.com/textmate/toml.tmbundle)
+- **TSQL:** [textmate/sql.tmbundle](https://github.com/textmate/sql.tmbundle)
```
There was a collision here when I attempted to add a different grammar for TSQL since sql.tmbundle already used up `source.sql` and the one I was adding also tried to add `source.sql`. If there is a way around that, I can try adding the grammar again.

lildude replied:
If a grammar supports multiple languages or variations of the language, as it appears this grammar might, then it's fine to reuse the same grammar.

lildude replied:
That said, I'm not 100% sure this grammar supports TSQL based on https://github.com/textmate/sql.tmbundle/issues/25 - it also appears to be unmaintained, so if you've got a better source that covers all the languages this grammar is currently used for, and it has support for TSQL, feel free to swap it in this PR.

beau-witter replied:
You are correct, and looking at files I know to be TSQL on GitHub, I can see some syntax highlighting issues. The problem I have is, I am not sure if there is a grammar that supports all of them. I checked the 2 you mentioned in the Issue, and the vscode one is at least still maintained, but looks like it works exclusively for TSQL, and the other one is just a little less stale than sql.tmbundle and likely only works for TSQL as well.

lildude replied:
If you can't find a grammar that covers all, the other option would be to submit a PR or issue to one of the still maintained grammars suggesting changing the scope to something like `source.tsql`.  https://github.com/kinnell/atom-language-transact-sql (I think this is a TSQL grammar) looks like it's quite well maintained so this might be a good place to start.

You can test the grammar using lightshow as referenced in the PR template.

beau-witter replied:
Lightshow does not appear to work well with the one you suggested (they use a .plist to defined the pattern, but use .less to manage coloring, I don't know if that would be a problem if it were to be used here). I will continue to poke around.

beau-witter replied:
This grammar (https://github.com/theia-ide/theia/blob/master/packages/textmate-grammars/data/sql.tmLanguage.json) seems to work really well for TSQL. Is it possible to add just that as the grammar? Or no?

lildude replied:
Nope. The grammar's in the wrong path for Linguist to pick it up. You can however use the original source - https://github.com/Microsoft/vscode-mssql/ - which would probably be a better choice as it's actively maintained. The only snag is it also uses the `source.sql` scope so will need to support all the other SQLs too 😞

beau-witter replied:
Darn, the reason I liked the Theia one better is because it actually works through Lightshow. The Microsoft one (https://github.com/Microsoft/vscode-mssql/blob/master/syntaxes/SQL.plist) does not.

lildude commented on lib/linguist/heuristics.yml:
```diff
@@ -383,13 +383,17 @@ disambiguations:
   - language: PLSQL
     pattern: '(?i:\$\$PLSQL_|XMLTYPE|sysdate|systimestamp|\.nextval|connect by|AUTHID (DEFINER|CURRENT_USER)|constructor\W+function)'
+  # T-SQL
+  - language: TSQL
```
The problems you experienced and mentioned at https://github.com/github/linguist/issues/4476#issuecomment-479202687 are caused by the fact your `pattern` and `negative_pattern` aren't being considered together. You need to "and" them like so:

```yaml
  # T-SQL
  - language: TSQL
    and:
      - negative_pattern: '(?i:IDENTIFIED|NUMBER|VARCHAR2)'
      - pattern: '(?i:(GO)|(CREATE PROCEDURE)|OUTPUT( INSERTED)|IF|ELSE|CURSOR|FETCH|DEALLOCATE|DECLARE)'
```

This will solve the problem you've encountered before, but will now cause a switch of classification for the `check_reorg.sql` file from SQLPL to TSQL so you're going to need a few more tweaks.

beau-witter replied:
I will make that update and then check out what other kind of tweaks I can add for remove the incorrect classification. Thanks!

beau-witter commented:
@lildude Updated the draft PR with some changes, all tests now pass. I think the major issue now is the grammar problem.

beau-witter commented:
@nanoquanta Yes, but I am not sure how to look. Do you know of something that may be able to help?

acdevick commented:
@nanoquanta Thanks for working on this our company just started using Flyway and moved all of our TSQL to github.  I came across this looking for a solution to fix temp tables where github is rendering #tempTable as a commented line vs actual TSQL Code.  Would that be able to be fixed as a feature to TSQL?

beau-witter commented:
@acdevick If we are able to get the right grammar file included as part of adding the TSQL language to Linguist, then yes, that should be rendered correctly assuming the file is correctly detected as TSQL.

beau-witter commented:
@lildude If I were to make a separate repo to be the TSQL grammar repo, have the scope be source.tsql, and have it setup in such a way that linguist can read/understand the path, would that be ok?

lildude commented:
@beau-witter yeah, I think so. You'd become the maintainer of that grammar going forward and can't delete it as long as Linguist continues to use it.

@Alhadis has more experience with maintaining grammars so should know if simply changing the scope is sufficient for our needs.

beau-witter commented:
@lildude I added the TSQL grammar repo I created as the grammar for the TSQL language I added to linguist. All the tests appear to pass. Let me know if there is anything else that needs to be done to move forward.

@Alhadis When I added https://github.com/beau-witter/language-tsql as a grammar using script/add-grammar it auto added the scope as `source.sql` when the tsql.json file I have in the grammars folder specifies the scope to be `source.tsql`. I changed it manually (observable in the most recent commit). I just want to check if that is ok, and if not, why it happened?

beau-witter commented:
I am turning this into a real PR as I believe it is ready to be fully reviewed.

lildude commented:
TIL... Travis doesn't support draft PRs yet so CI wasn't triggered when you flipped this PR to be a real PR. I'm going to see if I can trigger a build by closing and reopening the PR. If I hit problems, I may need you to do it @beau-witter.

lildude commented:
🎉 it worked!!!

beau-witter commented:
@lildude Alright! Thanks for Travis CI working! I think I got everything corrected on my side, and Travis seems to agree! Should be good to review now.

Alhadis commented:
Hey guys, sorry for the silence. Been out of action for the last couple of weeks. :wave:

> @Alhadis has more experience with maintaining grammars so should know if simply changing the scope is sufficient for our needs.

Yes, that's fine. A grammar's scope-name is mainly important to Atom for styling and syntax-level command scoping; e.g., an author might use `source.sql.tsql` to maximise compatibility with a user's customisations: a snippet that targets `source.sql` will also work for `source.sql.tsql`, but *not* `source.tsql`.

However, all of this is moot as far as Linguist is concerned, because a scope only has to be unique (in the sense of acting as a global identifier). I recommend using `source.tsql` because it's much clearer and obvious to future contributors.

beau-witter commented:
@Alhadis Thanks for that info! That is really neat. I did end up doing the `source.tsql` way, so thanks for confirming that way will work!

beau-witter commented:
@lildude Is there anything else I can do to help this PR move along?

lildude commented:
Nope. It's up to @pchaigno and/or @Alhadis to review next.

beau-witter commented:
@lildude I see the changes have been approved. Yay!

Just curious, is there a known update date when this might be merged by?

lildude commented:
> Just curious, is there a known update date when this might be merged by?

There's never a known date for merging as it's not really important.  I merge whenever I get around to reviewing, if I approve, but it won't take effect on GitHub.com until I make a release, which once against doesn't have a known date beyond "I aim for approximately once a month, based on my own load and the GitHub.com deploy queue and locks etc" 😄

esauser commented:
Any idea when this will be included in GitHub and/or GitHub Enterprise?

lildude commented:
I'm aiming to do a release this week. If I succeed, this will make it into GitHub Enterprise Server 2.17.1. If I can't get this on GitHub.com this week, it'll be in 2.17.2.
